### ISSUE TITLE
Mutate corresponding collections/documents

### ISSUE THREAD

**Issue by nandorojo** [nandorojo/swr-firestore#1]
# Why

No more Redux reducer jumble for dealing with globally-shared Firestore data. Enough is enough.

# What

If we `mutate` a certain document, say `users/fernando`, and we have an SWR query for the collection `users`, then we should also mutate that collection.

It would be great to detect in documents what their collection is, and mutate the document for that path. This means full in-sync Firestore data across the app.

The same should go for updating collection  -> documents. If you fetch / listen to a collection, and it updates, it should mutate every document with that path, with `shouldRevalidate` set to `false`.

# Roadblocks

This would all require that the `useSWR(key)`  key value is restricted to only the `path`. By including the `query` object here too, we can't perfectly sync data. The problem is that, if we don't update the `query` in the `key`, then changes to the `query` will not retrigger listeners/fetches, rendering the query stale. This is a fundamental problem to figure out. 

Does having a function key solve this?

**Update** I'm tackling this in a [subsequent comment](https://github.com/nandorojo/swr-firestore/issues/1#issuecomment-623187574).

# Suggestions

Might have to add an issue on SWR, which tells it that the `key` "changed", even though it's arguments have stayed the same. For instance, even though we might be listening to the same collection, we now have a different set of query options, so we want it to "re-validate". 

# Likely solution

Run a `useEffect` hook **after** `useSWR`, which listens for changes in `queryString`, `listen`, etc. And if it changes, run `revalidate`, returned by SWR. That should do it! Just make sure that the `shouldListen` ref and `query` ref are in a `useEffect` before `useSWR`.

**Comment by nandorojo** [comment]
`0.50.0` solves this for `collection` -> `document`. Any collection query will update all documents that it fetches from Firestore.

However, fetching a document doesn't necessarily fix the problem of mutating all collections that have queried that document. It's a bit unclear how to achieve this right now. The difficulty comes from the fact that a collection can have one path, but many different query objects. The query objects must be in the key of the collection, so I don't know how to tell the document query which collection `key` it should mutate.

# Solution

I think the solution here will be to have a `cache` that tracks all collection queries that get created for a given collection path. Every time `useCollection` gets called, it will add a field to this cache, telling it 1) the path and 2) the `query` string.

Next, the cache should export some sort of function that returns to you all the possible SWR `key` arguments that correspond to a given collection path. The document query will loop over these keys and mutate them.

If there are no queries in the cache for that collection path, the only item in the array will be the collection path itself. This will always be included. **(Scratch this point)**

However, if you've already made collection queries for that path, and also have query strings, those items will also be in the array.

The cache structure could look something like this:

```typescript
// cache.ts

type CollectionsCache = {
  [path: string]: {
    key: [string, string | undefined] // [path, queryString]
  }[]
}

class Cache {
  private collections: CollectionsCache

  constructor() {
    this.collections = {}
  }

  getSWRKeysForCollection(path: string) {
    return this.collections[path] ?? []
  }
  addCollectionToCache(path, queryString) {
   const { [path]: collection = [] } = this.collections
   this.collections = {
    ...this.collections,
    [path]: [
      ...collection,
      // TODO fix duplicate items getting added here. Should this be a Set? Dictionary?
      { key: [path, queryString] }
    ]
   }
    return this.collections
  }
}

export default new Cache()
```
_There is some data redundancy in the `CollectionsCache` type. That said, I think it's better to leave it this way to leave the door open for future API changes. That way, the API is always `getSWRKeysForCollection`, but the cache outline itself may change, from the way we generate SWR `key`s to what goes in the key._

Whenever you create a new collection query:

```js
import Cache from '../cache'

...

useSWR([path, queryString], (path, queryString) => {
  Cache.addCollectionToCache(path, queryString)
  ...
})
```

And when you load a new document:

```js
import Cache from '../cache'

...
const data = {
   ...doc.data()
   id: doc.id,
   // etc
}
const collectionPath = last(path.split('/'))

const promises = []

Cache.getSWRKeysForCollection(collectionPath).forEach(key => {
  promises.push(
    mutate(key, (state = []) => state.map(document => {
      if (document.id === data.id) return data
      return document
    }), false)
  )
})

return Promise.all(promises)
```
The one problem with the above solution is that it creates a new copy of the state array every time that collection is hit, even if the document isn't even in that collection.

The alternative is to loop through first to find that one document, and somehow only return a new array if that document is in the collection. At that point, we're comparing the performance of re-rendering vs. looping through. The looping might be the better move, but I'm not expert at this part.

**Comment by nandorojo** [comment]
This is now included in `0.7.x`, though the logic could be better, as it currently loops twice.
